### PR TITLE
Use references & flash strings where approperate

### PR DESCRIPTION
**Pull request by rjwats**

### CODE REVIEW

**Comment by rjwats** [comment]
@proddy I know you mentioned wanting to reduce pass by value in #79.

As part of this I've opted to move the largest strings to flash, probably the biggest memory gains to be found there. Holding off on looking at std::str for now... might come back to that.

Does this look like step in the right direction?

**Comment by proddy** [comment]
@rjwats thanks for implementing the pass by value. There aren't many large strings to put into Flash other than the Serial debug messages so not a huge win, but its always good practice. If many settings were actual strings it may make sense, although serializing with Json is costly as the ArduinoJson library will make a duplicate in RAM.

Still not a fan of the String class. I know it must be strange coming from Java where it just works, but Ardunio String's is notoriously evil and seriously chews up memory. You'll see a big difference in heap fragmentation when you switch to using 'const &std::string'. I did it last xmas on an earlier fork of esp8266-react with nice results. Also consider using StaticJsonDocument for those docs < 1kb to avoid using heap. I also don't think you need to create JsonObjects anymore and can pass the doc reference directly (new in ArduinoJson 6.15), although I would need to double-check!

**Comment by rjwats** [comment]
@kasedy do these changes seem sensible to you?

**Comment by kasedy** [comment]
I'll look into it a bit later.

File: lib/framework/JsonUtils.h [hunk 1]
```diff
@@ -7,14 +7,14 @@
...
-      _ip = INADDR_NONE;
+  static void readIP(JsonObject& root, const String& key, IPAddress& ip) {
+    if (!root[key].is<String>() || !ip.fromString(root[key].as<String>())) {
+      ip = INADDR_NONE;
     }
   }
-  static void writeIP(JsonObject& root, String key, IPAddress& _ip) {
-    if (_ip != INADDR_NONE) {
-      root[key] = _ip.toString();
+  static void writeIP(JsonObject& root, const String& key, IPAddress& ip) {
```
Comment (kasedy) [inline lib/framework/JsonUtils.h:15]:
```suggestion
  static void writeIP(JsonObject& root, const String& key, const IPAddress& ip) {
```

File: lib/framework/OTASettingsService.cpp [hunk 2]
```diff
@@ -39,27 +39,27 @@ void OTASettingsService::configureArduinoOTA() {
...
+    Serial.println(F("Starting OTA Update Service..."));
     _arduinoOTA = new ArduinoOTAClass;
     _arduinoOTA->setPort(_state.port);
     _arduinoOTA->setPassword(_state.password.c_str());
-    _arduinoOTA->onStart([]() { Serial.println("Starting"); });
-    _arduinoOTA->onEnd([]() { Serial.println("\nEnd"); });
+    _arduinoOTA->onStart([]() { Serial.println(F("Starting")); });
+    _arduinoOTA->onEnd([]() { Serial.println(F("\nEnd")); });
     _arduinoOTA->onProgress([](unsigned int progress, unsigned int total) {
       Serial.printf("Progress: %u%%\r", (progress / (total / 100)));
```
Comment (kasedy) [inline lib/framework/OTASettingsService.cpp:49]:
```suggestion
      Serial.printf_P(PSTR("Progress: %u%%\r"), (progress / (total / 100)));
```

File: lib/framework/WiFiStatus.cpp [hunk 3]
```diff
@@ -18,35 +18,35 @@ WiFiStatus::WiFiStatus(AsyncWebServer* server, SecurityManager* securityManager)
...
   Serial.println(info.disconnected.reason);
 }
 
 void WiFiStatus::onStationModeGotIP(WiFiEvent_t event, WiFiEventInfo_t info) {
-  Serial.print("WiFi Got IP. localIP=");
+  Serial.print(F("WiFi Got IP. localIP="));
   Serial.print(WiFi.localIP().toString());
-  Serial.print(", hostName=");
+  Serial.print(F(", hostName="));
   Serial.println(WiFi.getHostname());
```
Comment (kasedy) [inline lib/framework/WiFiStatus.cpp:33]: Serial.printf_P(PSTR(""WiFi Got IP. localIP=%s, hostName=%s\n"), WiFi.localIP().toString().c_str(), WiFi.getHostname().c_str());

printf is more readable.

File: src/LightStateService.cpp [hunk 4]
```diff
@@ -48,14 +48,14 @@ void LightStateService::registerConfig() {
...
   }
   String configTopic;
-  String setTopic;
-  String stateTopic;
+  String subTopic;
+  String pubTopic;
 
   DynamicJsonDocument doc(256);
   _lightMqttSettingsService->read([&](LightMqttSettings& settings) {
     configTopic = settings.mqttPath + "/config";
```
Comment (kasedy) [inline src/LightStateService.cpp:56]:
```suggestion
    configTopic = settings.mqttPath + F("/config");
```

**Comment by kasedy** [comment]
Wen we have a time it would be nice to improve logging system. Writing to serial is not bad. But it would be nice to have possibility:
1) completely disable logging at compile time
2) Redirect logging to websocket, udp, mqtt, etc.
3) Have several levels of logging (TRACE, INFO, WARNING, ERROR)
4) Enable/disable logging for a class/module.

**Comment by kasedy** [comment]
1) I like the changes as passing variable by reference is always faster and takes less space on the stack. I think there are other places where we can use reference.
2) I always put strings into PROGMEM if the string is not used during whole life of the program. RAM is much smaller than flash. it is better to spend time on copying string from FLASH to RAM than bump into end of memory.
3) I prefer to use String over std::string as it is Arduino/ESP common used container. The use of two classes simultaneously decreases readability. 
4) I prefer String over const char* unless I low on memory or flash. The use of const char* make code more complicated as you need to track its length and memory allocations.

**Comment by rjwats** [comment]
Thanks for the feedback on the PR - I've updated it and tested the changes.

I agree with your comments on the logging - I have a logging abstraction in a branch somwhere which never made it's way to master. I'm not a massive fan of working with char pointers either. I guess there are sometimes memory or performance benefits for doing it that way but I find the String approach much more readable. Code I wrote in ArduinoJsonJWT is a good example of where it might be possible to improve the performance, but probably at the cost of readability.

**Comment by proddy** [comment]
Agree on @kasedy's points 1-4 (although I still think String is evil, see https://github.com/rjwats/esp8266-react/issues/79). For logging I did a test back in https://github.com/rjwats/esp8266-react/issues/85 using different log levels sent to a console and/or an external syslog daemon and it works well.

I love this project but holding back updating my projects until the master is up to date with all the great feature branches. That's because I would need to make many small changes, for example removing all the Serial.* calls.

**Pull request merged**